plugins/inventory/foreman.py

```diff
@@ -238,10 +297,200 @@ def _get_hostvars(self, host, vars_prefix='', omitted_vars=()):
+        params["IPv4"] = options[self.want_IPv4]
+        params["IPv6"] = options[self.want_IPv6]
+        params["Facts"] = options[self.want_facts]
+        params["Host Group"] = options[self.want_host_group]
+        params["Host Collections"] = options[self.want_hostcollections]
+        params["Subnet"] = options[self.want_subnet]
+        params["Subnet v6"] = options[self.want_subnet_v6]
+        params["Smart Proxies"] = options[self.want_smart_proxies]
+        params["Content Attributes"] = options[self.want_content_facet_attributes]
+        params["Host Parameters"] = options[self.want_host_params]
+        if self.host_filters:
+            params["Hosts"] = self.host_filters
+        return params
+
+    def _use_inventory_report(self):
+        use_inventory_report = self.get_option('foreman').get('use_reports_api')
+        if not use_inventory_report:
+            return False
+        status_url = "%s/api/v2/status" % self.foreman_url
+        result = self._get_json(status_url)
+        foreman_version = (LooseVersion(result.get('version')) >= LooseVersion(self.MINIMUM_FOREMAN_VERSION_FOR_REPORTING_API))
+        return foreman_version
+
+    def _post_request(self):
+        url = "%s/ansible/api/v2/ansible_inventories/schedule" % self.foreman_url
+        session = self._get_session()
+        params = {'input_values': self._fetch_params()}
+        self.poll_interval = self.get_option('report').get('poll_interval')
+        ret = session.post(url, json=params)
+        if not ret:
+            raise Exception("Error scheduling inventory report on foreman. Please check foreman logs!")
+        url = "{0}/{1}".format(self.foreman_url, ret.json().get('data_url'))
+        response = session.get(url)
+        while response:
+            if response.status_code != 204:
+                break
+            sleep(self.poll_interval)
+            response = session.get(url)
+        if not response:
+            raise Exception("Error receiving inventory report from foreman. Please check foreman logs!")
+        else:
+            return response.json()
+
     def _populate(self):
+        if self._use_inventory_report():
+            self._populate_report_api()
+        else:
+            self._populate_host_api()
+
+    def _populate_report_api(self):
+        self.groups = dict()
+        self.hosts = dict()
+        try:
+            inventory_report_response = self._post_request()
+        except Exception:
+            self._populate_host_api()
+            return
+        self.group_prefix = self.get_option('group_prefix')
+        host_data = json.loads(inventory_report_response)
+        for host in host_data:
+            if not(host) or (host["name"] in self.cache.keys()):
+                continue
+            dns_name = self.inventory.add_host(host['name'])
+            host_name = dns_name
+            group_name = host.get('hostgroup_title', host.get('hostgroup_name'))
```

Comment:
@ezr-ondrej has recently fixed a bug where we had accidentally omitted empty parent groups from the import, is the same needed in this part too? see https://github.com/theforeman/foreman-ansible-modules/pull/931

Reply:
Updated and tested with regular hostgroup with hosts. Could you explain the exact data setup I should have to test this.

Reply:
Have a look at the test playbook changes in #931. Essentially: create a groupA, create a groupB with parent=groupA, now put a host into groupB, the inventory should report it as being in ansbibke group groupA/groupB (or whatever safe does from it, I think groupA_groupB).

Technically it should also be in groupA, but I think that's not implemented in API mode either. Right @ezr-ondrej?

Reply:
Ok..I tested with parent/child hostgroup and I see the same behavior with reporting api and host api.
The data looks like this with both: (Empty parent hostgroup: "test_host_group", Child Hostgroup: "test_group_child")

"foreman_test_host_group": {
        "children": [
            "foreman_test_host_group_test_group_child"
        ]
    }, 
    "foreman_test_host_group_test_group_child": {
        "hosts": [
            "virt-who-0e93693d.7096.4346.afb6.1c95bd1153e9-1"
        ]
    },

Reply:
(in groupA of course works, that's what Ansible does for us)

Reply:
Thanks! I had the tests a bit refactored last week and will try to add them to this PR next week when back home, shouldn't be too hard 🤘